cpp/src/arrow/compute/row/row_internal.h

```diff
@@ -0,0 +1,250 @@
+  ///     data.  This is ignored if there are no variable-length columns
+  Status AppendEmpty(uint32_t num_rows_to_append, uint32_t num_extra_bytes_to_append);
+  /// \brief Append rows from a source table
+  /// \param from The table to append from
+  /// \param num_rows_to_append The number of rows to append
+  /// \param source_row_ids Indices (into `from`) of the desired rows
+  Status AppendSelectionFrom(const RowTableImpl& from, uint32_t num_rows_to_append,
+                             const uint16_t* source_row_ids);
+  /// \brief Metadata describing the data stored in this table
+  const RowTableMetadata& metadata() const { return metadata_; }
+  /// \brief The number of rows stored in the table
+  int64_t length() const { return num_rows_; }
+  // Accessors into the table's buffers
+  const uint8_t* data(int i) const {
+    ARROW_DCHECK(i >= 0 && i < kMaxBuffers);
+    return buffers_[i];
+  }
+  uint8_t* mutable_data(int i) {
+    ARROW_DCHECK(i >= 0 && i < kMaxBuffers);
+    return mutable_buffers_[i];
+  }
+  const uint32_t* offsets() const { return reinterpret_cast<const uint32_t*>(data(1)); }
+  uint32_t* mutable_offsets() { return reinterpret_cast<uint32_t*>(mutable_data(1)); }
+  const uint8_t* null_masks() const { return null_masks_->data(); }
+  uint8_t* null_masks() { return null_masks_->mutable_data(); }
+
+  /// \brief True if there is a null value anywhere in the table
+  ///
+  /// This calculation is memoized based on the number of rows and assumes
+  /// that values are only appended (and not modified in place) between
+  /// successive calls
+  bool has_any_nulls(const LightContext* ctx) const;
+
+ private:
+  Status ResizeFixedLengthBuffers(int64_t num_extra_rows);
+  Status ResizeOptionalVaryingLengthBuffer(int64_t num_extra_bytes);
+
+  // Helper functions to determine the number of bytes needed for each
+  // buffer given a number of rows.
+  int64_t size_null_masks(int64_t num_rows);
+  int64_t size_offsets(int64_t num_rows);
+  int64_t size_rows_fixed_length(int64_t num_rows);
+  int64_t size_rows_varying_length(int64_t num_bytes);
+  // Called after resize to fix pointers
+  void update_buffer_pointers();
+
+  // The arrays in `buffers_` need to be padded so that
+  // vectorized operations can operate in blocks without
+  // worrying about tails
+  static constexpr int64_t kPaddingForVectors = 64;
+  MemoryPool* pool_;
+  RowTableMetadata metadata_;
+  // Buffers can only expand during lifetime and never shrink.
+  std::unique_ptr<ResizableBuffer> null_masks_;
+  // Only used if the table has variable-length columns
+  // Stores the offsets into the binary data (which is stored
+  // after all the fixed-sized fields)
+  std::unique_ptr<ResizableBuffer> offsets_;
+  // Stores the fixed-length parts of the rows
+  std::unique_ptr<ResizableBuffer> rows_;
+  static constexpr int kMaxBuffers = 3;
+  const uint8_t* buffers_[kMaxBuffers];
+  uint8_t* mutable_buffers_[kMaxBuffers];
```

Comment:
Are these different except for the different const-ness?

Reply:
Nope. I'm removing the const version to make things simpler